### PR TITLE
Convert samples to use shared_ptr stream creation interface

### PR DESCRIPTION
**Pull request by ggfan**
This is for github issue https://github.com/google/oboe/issues/830, please kindly take a look, thank you!

### CODE REVIEW

File: samples/RhythmGame/src/main/cpp/Game.cpp [hunk 1]
```diff
@@ -56,10 +56,8 @@ void Game::start() {
 
 void Game::stop(){
 
-    if (mAudioStream != nullptr){
+    if (mAudioStream){
```
Comment (philburk) [inline samples/RhythmGame/src/main/cpp/Game.cpp:59]: To be safe, call stop() before close().

File: samples/iolib/src/main/cpp/player/SimpleMultiPlayer.cpp [hunk 2]
```diff
@@ -135,6 +135,7 @@ void SimpleMultiPlayer::teardownAudioStream() {
     // tear down the player
     if (mAudioStream != nullptr) {
         mAudioStream->stop();
+        mAudioStream->close();
```
Comment (philburk) [inline samples/iolib/src/main/cpp/player/SimpleMultiPlayer.cpp:138]: mAudioStream.reset();

**Review by philburk — CHANGES_REQUESTED** [review]

File: samples/RhythmGame/src/main/cpp/Game.cpp [hunk 3]
```diff
@@ -56,10 +56,8 @@ void Game::start() {
...
         mAudioStream->close();
-        delete mAudioStream;
-        mAudioStream = nullptr;
     }
```
[the remark this answers is not in the document]
Comment (ggfan) [reply samples/RhythmGame/src/main/cpp/Game.cpp:61]: added.

File: samples/MegaDrone/src/main/cpp/MegaDroneEngine.h [hunk 4]
```diff
@@ -41,7 +41,7 @@ class MegaDroneEngine : public IRestartable {
     virtual void restart() override;
 
 private:
-    oboe::ManagedStream mStream;
+    std::shared_ptr<AudioStream> mStream;
```
[the remark this answers is not in the document]
Comment (ggfan) [reply samples/MegaDrone/src/main/cpp/MegaDroneEngine.h:44]: It seems that restart() is only called from ErrorAfterClose() callback, in which case the stream is already closed., so restart() could assume that stream is in closed state.

the above does not solve the destructor case; wondering we could do: 
  - destructor: stop and close stream if it is NOT closed yet; ( we do not have to call mStream->reset() here as it is going away
 - that mirrors the creation inside constructor.

WDYT?

File: samples/RhythmGame/src/main/cpp/Game.cpp (hunk 1 above)
Comment (ggfan) [reply samples/RhythmGame/src/main/cpp/Game.cpp:59]: done

File: samples/iolib/src/main/cpp/player/SimpleMultiPlayer.cpp (hunk 2 above)
Comment (ggfan) [reply samples/iolib/src/main/cpp/player/SimpleMultiPlayer.cpp:138]: done

File: samples/MegaDrone/src/main/cpp/MegaDroneEngine.h (hunk 4 above)
[the remark this answers is not in the document]
Comment (dturner) [reply samples/MegaDrone/src/main/cpp/MegaDroneEngine.h:44]: @philburk Why shouldn't `MegaDroneEngine` call `start()` in its constructor? It doesn't make much sense (to me) to have the engine object without a running playback stream.

File: samples/LiveEffect/src/main/cpp/LiveEffectEngine.h [hunk 5]
```diff
@@ -23,6 +23,8 @@
 #include <thread>
 #include "FullDuplexPass.h"
 
+using SharedStream = std::shared_ptr<oboe::AudioStream>;
```
[the remark this answers is not in the document]
Comment (ggfan) [reply samples/LiveEffect/src/main/cpp/LiveEffectEngine.h]: done

**Review by dturner — APPROVED** [review]
LGTM but it'd be good to get Phil's view on the outstanding question around MegaDrone starting a stream inside the constructor.

File: samples/MegaDrone/src/main/cpp/MegaDroneEngine.h [hunk 6]
```diff
@@ -33,21 +33,22 @@ class MegaDroneEngine : public IRestartable {
...
     std::shared_ptr<TappableAudioSource> mAudioSource;
     std::unique_ptr<DefaultAudioStreamCallback> mCallback;
 
     oboe::Result createPlaybackStream();
     void createCallback(std::vector<int> cpuIds);
     void start();
+    void stop(void);
```
Comment (philburk) [inline samples/MegaDrone/src/main/cpp/MegaDroneEngine.h:51]: void not needed

**Review by philburk — CHANGES_REQUESTED** [review]

**Review by philburk — APPROVED** [review]

**Pull request merged**